Algorithms/Dynamic-Programming/Travelling Salesman Problem.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+This program addresses the famous, Travelling Salesman Problem:
+Probem Statement: Given a set of cities and distance between every pair of cities, 
+				  the problem is to find the shortest possible route that visits every 
+				  city exactly once and returns to the starting point.
+Input Format : The first line contains n,the number of cities in the region. (1 <= n <=20 (Explained Below))
+			   Each of the next n lines contain n numbers describing the distance matrix.
+			   (The distance from a city to itself is taken as zero)
+Output Format: A single number , the length of the shortest such path.
+
+Approach Used : Dynamic Programming, Bit Masking
+
+Intuition for using Bit masking:
+The bitmasking technique is used to keep track of the cities visited so far.
+Say for example there are 4 cities and after certain steps cities B,D are visited only.
+At this instant the variable mask would appear as 1010 as the MSB signifies that city D has been visited
+and the LSB signifies that city A is not yet visited and so on. Now suppose if we carry out an AND operation as
+1010 & 0010(for City B) , the result is non zero if City B has been already visited and zero if City B
+is not yet visited(1000 & 0010). This approach is used to continue the recursion. Bit masking + the current city position
+uniquely specifies the subproblem for the main-problem. 
+
+
+
+Time Complexity Analysis:
+It is clear that at each stage of the problem there can be a maximum n * (2^n) subproblems, with each taking O(n) time to compute.
+So the overall time complexity is O( (n*n) * (2^n) ) . This is very less than that of the brute approach i.e O(n!). However this
+approach is not suitable for very high values of n either. It is due to the size constraint of the 2d dp array. The maximum number of
+cities can be taken 22 . So 20 can be taken as a safe limit. (Maximum global array size : 10^9).
+
+
+Author : monsijb
+*/
+
+#include<bits/stdc++.h>
+using namespace std;
+
+int n;
+int dp[1048576][20];
+int dist[20][20];
+int visited_all;
+
+// A function to calculate (base)^(exponent) noting that pow gives error at high values.
+int power(int base,int exp)
+{
+	int result = 1;
+	for(int i=1; i<=exp; i++)
+	{
+		result = result * base;
+	}
+	return result;
+}
+
+// Function to initialize dp array
+void init(int n)
+{
+	for(int i=0; i<(1<<n); i++)
+	{
+		for(int j=0; j<n; j++)
+		{
+			dp[i][j]=-1;
+		}
+	}
+}
+
+// Recursive function to calculate the final result
+int tsproblem(int mask,int position)
```

Comment:
The function name should be `tsProblem()` according to [Coding standard](https://github.com/sahilbansal17/Competitive_Coding/wiki/Coding-Standards#naming-convention)